File: python_src/11.盛最多水的容器.py
#
# @lc app=leetcode.cn id=11 lang=python3
#
# [11] 盛最多水的容器
#
from typing import *

# @lc code=start
class Solution:
    def maxArea(self, height: List[int]) -> int:
        size = len(height)
        l, r = 0, size - 1
        # l0, r0 = 0, size
        lmax, rmax = height[l], height[r]
        cap = 0
        while l < r:
            h = lmax if lmax < rmax else rmax
            cap0 = h * (r - l)
            cap = cap if cap > cap0 else cap0
            if height[l] < height[r]:
                l += 1
                # lmax = max(lmax, height[l])
                lmax = lmax if lmax > height[l] else height[l]
            else:
                r -= 1
                # rmax = max(rmax, height[r])
                rmax = rmax if rmax > height[r] else height[r]

        return cap

# @lc code=end

s = Solution()
print(s.maxArea([1,8,6,2,5,4,8,3,7]))
print(s.maxArea([1,1]))

File: c_src/11.盛最多水的容器.c
/*
 * @lc app=leetcode.cn id=11 lang=c
 *
 * [11] 盛最多水的容器
 */
#include <math.h>
#include "print_utils.h"

// @lc code=start
#define max(a, b) ((a) > (b) ? (a) : (b))
#define min(a, b) ((a) < (b) ? (a) : (b))
int maxArea(int* height, int heightSize) {
    size_t l = 0, r = heightSize - 1;
    int area = min(height[l], height[r]) * (r - l);

    while (l < r) {
        if (height[l] < height[r]) {
            l++;
        } else {
            r--;
        }
        area = max(area, (r - l) * min(height[r], height[l]));
    }
    return area;
}
// @lc code=end

void test() {
    static int arr[] = { 1,8,6,2,5,4,8,3,7 };
    print_int(maxArea(arr, 9));

    static int arr2[] = { 1,1 };
    print_int(maxArea(arr2, 2));
}